### PR TITLE
custom alerting tuning

### PR DESCRIPTION
**Pull request by junotx**
**What type of PR is this?**

/kind optimization

**What this PR does / why we need it**:

some optimizations for custom alerting

### CODE REVIEW

File: pkg/models/alerting/rules/ruler.go [hunk 1]
```diff
@@ -403,57 +415,77 @@ func (r *ThanosRuler) UpdateAlertingRule(ctx context.Context, ruleNamespace *cor
...
+		found        bool
+		success      bool
+		prsToDelRule = make(map[string]*promresourcesv1.PrometheusRule)
 	)
-	for _, prometheusRule := range prometheusRules {
-		resource := ruleResource(*prometheusRule)
+	for i, prometheusRule := range prometheusRules {
 		if success { // If the update has been successful, delete the possible same rule in other resources
-			if ok, err := resource.deleteAlertingRule(rule.Alert); err != nil {
+			if err := r.doRuleResourceOperation(prometheusRule, func(newerPr *promresourcesv1.PrometheusRule) error {
```
Comment (benjaminhuo) [inline pkg/models/alerting/rules/ruler.go:424]: all `newerPr` could be changed to `pr`

**Review by benjaminhuo — CHANGES_REQUESTED** [review]
This PR looks great! 
Just some minor changes are needed.

**Review by benjaminhuo — APPROVED** [review]
/lgtm
@zryfish Would you take a look at this PR?

**Review by benjaminhuo — APPROVED** [review]
/lgtm

File: pkg/api/alerting/v2alpha1/types.go [hunk 2]
```diff
@@ -26,21 +26,27 @@ import (
...
 	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/template"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
 const (
 	RuleLevelCluster   RuleLevel = "cluster"
 	RuleLevelNamespace RuleLevel = "namespace"
+
+	AnnotationKeyRuleUpdateTime = "rule_update_time"
```
Comment (zryfish) [inline pkg/api/alerting/v2alpha1/types.go:39]: normally, annotation key format follows `rule-update-time`, using slash instead of underscode

Comment (junotx) [reply pkg/api/alerting/v2alpha1/types.go:39]: This annotation key is intended for use within prometheus, so following its common usage

Comment (benjaminhuo) [reply pkg/api/alerting/v2alpha1/types.go:39]: here is an example of an annotation used in the alerts Prometheus sent:
https://github.com/kubesphere/notification-manager/blob/master/pkg/webhook/testdata/alert.json#L17

**Pull request merged**